doc/metaprogramming.md

```diff
@@ -1 +1,139 @@
 # Metaprogramming
+
+Macros are a language feature that allows writing code that writes other code (metaprogramming).
+
+In Lean 4, macros are used pervasively. So much so that core language features such as `do` notation
+is implemented via macros! As a language user, macros are useful to easily
+embed domain-specific languages and to generate code at compile-time, to name a few uses.
+
+
+#### Balanced Parentheses as an Embedded Domain Specific Language
+
+Let's look at how to use macros to extend the Lean 4 parser and embed a language for building _balanced parentheses_.
+This language accepts strings given by the [BNF grammar](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form)
+
+```
+Dyck := 
+  "(" Dyck ")" 
+  | "{" Dyck '}'
+  | "End"
+```
+
+We begin by defining an inductive data type of the grammar we wish to parse:
+
+```lean
+inductive Dyck: Type :=
+   | Round : Dyck -> Dyck -- ( <inner> )
+   | Flower : Dyck -> Dyck -- { <inner> }
```

Comment:
I don't think I've ever heard of flower brackets :) . Sounds very poetic though. Is it a common name?

Reply:
That's what I was taught as a child. It appears to be a colloquialism :) I'll be happy to change it to curly brackets if you'd like!

Reply:
Yeah, feel free to change this file in the new PR so we have both files side by side